File: project/inung-onboarding/src/test/java/icd/onboarding/surveyproject/survey/controller/SurveyControllerTest.java
package icd.onboarding.surveyproject.survey.controller;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import icd.onboarding.surveyproject.common.NotImplementedTestException;
import icd.onboarding.surveyproject.survey.common.controller.GlobalExceptionHandler;
import icd.onboarding.surveyproject.survey.controller.consts.ErrorCodes;
import icd.onboarding.surveyproject.survey.controller.dto.*;
import icd.onboarding.surveyproject.survey.controller.exception.CommonSurveyHttpException;
import icd.onboarding.surveyproject.survey.fixtures.ResponseFixtures;
import icd.onboarding.surveyproject.survey.fixtures.SurveyFixtures;
import icd.onboarding.surveyproject.survey.service.SurveyService;
import icd.onboarding.surveyproject.survey.service.domain.Survey;
import icd.onboarding.surveyproject.survey.service.exception.InSufficientOptionException;
import icd.onboarding.surveyproject.survey.service.exception.SurveyNotFoundException;
import icd.onboarding.surveyproject.survey.service.exception.UnsupportedOptionException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

import java.util.List;
import java.util.UUID;
import java.util.stream.IntStream;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest({SurveyController.class, GlobalExceptionHandler.class})
class SurveyControllerTest {
	@MockitoBean
	SurveyService surveyService;

	@Autowired
	MockMvc mockMvc;

	@Autowired
	ObjectMapper objectMapper;

	@DisplayName("설문 조회")
	@Nested
	class GetSurvey {
		final UUID surveyId = UUID.randomUUID();
		final int version = 1;

		@Test
		@DisplayName("조회 대상 설문 조사가 없으면 404 status와 error를 반환한다.")
		void throwExceptionWhenSurveyNotFound () throws Exception {
			// given
			given(surveyService.findByIdAndVersion(surveyId, version))
					.willThrow(new SurveyNotFoundException());

			// when & then
			mockMvc.perform(get("/api/v1/survey/{surveyId}/version/{version}", surveyId, version))
				   .andExpect(status().isNotFound())
				   .andExpect(jsonPath("$.error.message").value(ErrorCodes.SURVEY_NOT_FOUND.message))
				   .andExpect(jsonPath("$.error.code").value(ErrorCodes.SURVEY_NOT_FOUND.name()));
		}

		@Test
		@DisplayName("정보가 올바른 경우 200 status와 data를 반환한다.")
		void shouldGetSurveys () throws Exception {
			// given
			Survey survey = SurveyFixtures.basicSurvey();
			SurveyResponse response = SurveyResponse.fromDomain(survey);

			given(surveyService.findByIdAndVersion(surveyId, version))
					.willReturn(survey);

			// when & then
			mockMvc.perform(get("/api/v1/survey/{surveyId}/version/{version}", surveyId, version))
				   .andExpect(status().isOk())
				   .andExpect(jsonPath("$.data.id").value(response.id().toString()))
				   .andExpect(jsonPath("$.data.title").value(response.title()))
				   .andExpect(jsonPath("$.data.questions").isNotEmpty());
		}
	}

	@DisplayName("설문 생성")
	@Nested
	class CreateSurvey {
		final String url = "/api/v1/survey";

		@ParameterizedTest
		@NullAndEmptySource
		@DisplayName("설문 조사의 제목이 존재하지 않으면 400 status와 error를 반환한다.")
		void throwExceptionWhenSurveyNameIsNullAndEmpty (String title) throws Exception {
			// given
			CreateSurveyRequest request = new CreateSurveyRequest(title, "", List.of());

			given(surveyService.createSurvey(request.toDomain()))
					.willReturn(null);

			// when & then
			mockMvc.perform(post(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest())
				   .andReturn();

			verify(surveyService, never()).createSurvey(any());
		}

		@Test
		@DisplayName("질문이 존재하지 않으면 400 status와 error를 반환 한다.")
		void throwExceptionWhenInSufficientQuestions () throws Exception {
			// given
			CreateSurveyRequest request = new CreateSurveyRequest("설문 조사 1", "", List.of());

			given(surveyService.createSurvey(request.toDomain()))
					.willReturn(null);

			// when & then
			mockMvc.perform(post(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest())
				   .andReturn();

			verify(surveyService, never()).createSurvey(any());
		}

		@Test
		@DisplayName("질문이 10개 이상이면 400 status와 error를 반환한다.")
		void throwExceptionWhenMaxExceededQuestions () throws Exception {
			// given
			List<CreateQuestionRequest> questions = IntStream.range(1, 12)
															 .mapToObj(i -> new CreateQuestionRequest(
																	 "질문 " + i,
																	 "질문 설명 " + i,
																	 "LONG_TEXT",
																	 false,
																	 i,
																	 List.of())
															 ).toList();
			CreateSurveyRequest request = new CreateSurveyRequest("설문 조사 1", "", questions);

			given(surveyService.createSurvey(request.toDomain()))
					.willReturn(null);

			// when & then
			mockMvc.perform(post(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest())
				   .andReturn();

			verify(surveyService, never()).createSurvey(any());
		}

		@Test
		@DisplayName("질문의 유형이 올바르지 않으면 400 status와 error를 반환한다.")
		void throwExceptionWhenWrongInputType () throws Exception {
			// given
			CreateQuestionRequest question = new CreateQuestionRequest(
					"질문 1",
					"질문 설명",
					"INVALID_TYPE",
					true,
					1,
					List.of()
			);
			CreateSurveyRequest request = new CreateSurveyRequest("설문 조사 1", "", List.of(question));

			given(surveyService.createSurvey(any()))
					.willReturn(null);

			// when & then
			mockMvc.perform(post(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest())
				   .andReturn();

			verify(surveyService, never()).createSurvey(any());
		}

		@Test
		@DisplayName("질문의 유형이 선택형일 경우 옵션이 존재하지 않으면 400 status와 error를 반환한다.")
		void throwExceptionWhenOptionsNotHaveForSelectTypeQuestion () throws Exception {
			// given
			CreateQuestionRequest question = new CreateQuestionRequest(
					"질문 1",
					"질문 설명",
					"SINGLE_SELECT",
					true,
					1,
					List.of()
			);

			CreateSurveyRequest request = new CreateSurveyRequest(
					"설문 조사 1",
					"",
					List.of(question)
			);

			given(surveyService.createSurvey(any()))
					.willThrow(new InSufficientOptionException());

			// when & then
			mockMvc.perform(post(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest())
				   .andExpect(jsonPath("$.error.message").value(ErrorCodes.INSUFFICIENT_OPTION.message))
				   .andExpect(jsonPath("$.error.code").value(ErrorCodes.INSUFFICIENT_OPTION.name()))
				   .andReturn();

			verify(surveyService, never()).createSurvey(any());
		}

		@Test
		@DisplayName("질문의 유형이 문자형일 경우 옵션이 존재하면 400 status와 error를 반환한다.")
		void throwExceptionWhenOptionsExistsToTextTypeQuestion () throws Exception {
			// given
			CreateQuestionRequest questionRequest = new CreateQuestionRequest(
					"질문 1",
					"질문 설명",
					"SHORT_TEXT",
					false,
					1,
					List.of(new CreateOptionRequest("옵션 1", 1))
			);
			CreateSurveyRequest request = new CreateSurveyRequest("설문 조사1", "설문 조사 설명", List.of(questionRequest));

			given(surveyService.createSurvey(any()))
					.willThrow(new UnsupportedOptionException());

			// when & then
			mockMvc.perform(post(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest())
				   .andExpect(jsonPath("$.error.message").value(ErrorCodes.UNSUPPORTED_OPTION.message))
				   .andExpect(jsonPath("$.error.code").value(ErrorCodes.UNSUPPORTED_OPTION.name()))
				   .andReturn();

			verify(surveyService, never()).createSurvey(any());
		}

		@Test
		@DisplayName("설문 조사의 정보가 올바른 경우 200 stauts와 data를 반환한다.")
		void shouldCreateSurveyWithValidInfo () throws Exception {
			// given
			CreateQuestionRequest questionRequest = new CreateQuestionRequest(
					"질문 1",
					"질문 설명",
					"SHORT_TEXT",
					false,
					1,
					List.of()
			);
			CreateSurveyRequest request = new CreateSurveyRequest("설문 조사1", "설문 조사 설명", List.of(questionRequest));
			Survey survey = SurveyFixtures.basicSurvey();

			given(surveyService.createSurvey(any())).willReturn(survey);

			// when & then
			mockMvc.perform(post(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isCreated())
				   .andExpect(jsonPath("$.data.id").value(survey.getId().toString()));

			verify(surveyService, times(1)).createSurvey(any());
		}
	}

	@DisplayName("설문 수정")
	@Nested
	class UpdateSurvey {
		final String url = "/api/v1/survey";
		final UUID surveyId = UUID.randomUUID();

		@Test
		@DisplayName("설문이 존재하지 않으면 404 status와 error를 반환한다.")
		void throwExceptionWhenSurveyNotFound () throws Exception {
			UpdateQuestionRequest questionRequest = new UpdateQuestionRequest(
					"수정 질문 1",
					"",
					"SHORT_TEXT",
					1,
					false,
					List.of()
			);
			UpdateSurveyRequest request = new UpdateSurveyRequest(surveyId, 1, "수정된 설문", "수정된 설명", List.of(questionRequest));

			given(surveyService.updateSurvey(any()))
					.willThrow(new SurveyNotFoundException());

			mockMvc.perform(put(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isNotFound())
				   .andExpect(jsonPath("$.error.code").value(ErrorCodes.SURVEY_NOT_FOUND.name()))
				   .andExpect(jsonPath("$.error.message").value(ErrorCodes.SURVEY_NOT_FOUND.message));

			verify(surveyService, times(1)).updateSurvey(any());
		}

		@Test
		@DisplayName("질문이 존재하지 않으면 400 status와 error를 반환 한다.")
		void throwExceptionWhenInSufficientQuestions () throws Exception {
			UpdateSurveyRequest request = new UpdateSurveyRequest(surveyId, 1, "수정된 설문", "수정된 설명", List.of());

			mockMvc.perform(put(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest());

			verify(surveyService, never()).updateSurvey(any());
		}

		@Test
		@DisplayName("질문이 10개 이상이면 400 status와 error를 반환한다.")
		void throwExceptionWhenMaxExceededQuestions () throws Exception {
			List<UpdateQuestionRequest> questions = IntStream.rangeClosed(1, 11)
															 .mapToObj(i -> new UpdateQuestionRequest(
																	 "질문 " + i,
																	 "설명 " + i,
																	 "SHORT_TEXT",
																	 i,
																	 false,
																	 List.of()
															 )).toList();

			UpdateSurveyRequest request = new UpdateSurveyRequest(surveyId, 1, "설문 수정", "내용", questions);

			mockMvc.perform(put(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest());

			verify(surveyService, never()).updateSurvey(any());
		}

		@Test
		@DisplayName("질문의 유형이 올바르지 않으면 400 status와 error를 반환한다.")
		void throwExceptionWhenWrongInputType () throws Exception {
			UpdateQuestionRequest question = new UpdateQuestionRequest(
					"질문", "설명", "INVALID_TYPE", 1, true, List.of()
			);
			UpdateSurveyRequest request = new UpdateSurveyRequest(surveyId, 1, "설문 수정", "내용", List.of(question));

			mockMvc.perform(put(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest());

			verify(surveyService, never()).updateSurvey(any());
		}

		@Test
		@DisplayName("질문의 유형이 선택형일 경우 옵션이 존재하지 않으면 400 status와 error를 반환한다.")
		void throwExceptionWhenOptionsNotHaveForSelectTypeQuestion () throws Exception {
			UpdateQuestionRequest question = new UpdateQuestionRequest(
					"질문", "설명", "SINGLE_SELECT", 1, true, List.of()
			);
			UpdateSurveyRequest request = new UpdateSurveyRequest(surveyId, 1, "설문 수정", "내용", List.of(question));

			given(surveyService.updateSurvey(any()))
					.willThrow(new InSufficientOptionException());

			mockMvc.perform(put(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest())
				   .andExpect(jsonPath("$.error.code").value(ErrorCodes.INSUFFICIENT_OPTION.name()))
				   .andExpect(jsonPath("$.error.message").value(ErrorCodes.INSUFFICIENT_OPTION.message));

			verify(surveyService, never()).updateSurvey(any());
		}

		@Test
		@DisplayName("질문의 유형이 문자형일 경우 옵션이 존재하면 400 status와 error를 반환한다.")
		void throwExceptionWhenOptionsExistsToTextTypeQuestion () throws Exception {
			UpdateQuestionRequest question = new UpdateQuestionRequest(
					"질문",
					"설명",
					"SHORT_TEXT",
					1,
					false,
					List.of(new UpdateOptionRequest("옵션 1", 1))
			);
			UpdateSurveyRequest request = new UpdateSurveyRequest(surveyId, 1, "설문 수정", "내용", List.of(question));

			given(surveyService.updateSurvey(any()))
					.willThrow(new UnsupportedOptionException());

			mockMvc.perform(put(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isBadRequest())
				   .andExpect(jsonPath("$.error.code").value(ErrorCodes.UNSUPPORTED_OPTION.name()))
				   .andExpect(jsonPath("$.error.message").value(ErrorCodes.UNSUPPORTED_OPTION.message));

			verify(surveyService, never()).updateSurvey(any());
		}

		@Test
		@DisplayName("수정 요청에 대한 정보가 올바른 경우 200 status와 data를 반환한다.")
		void shouldUpdateSurveyWithValidInfo () throws Exception {
			UpdateQuestionRequest question = new UpdateQuestionRequest(
					"질문", "설명", "SHORT_TEXT", 1, false, List.of()
			);
			UpdateSurveyRequest request = new UpdateSurveyRequest(surveyId, 1, "설문 수정", "내용", List.of(question));
			Survey updatedSurvey = SurveyFixtures.updatedSurvey(request.id(), request.version());

			given(surveyService.updateSurvey(any())).willReturn(updatedSurvey);

			mockMvc.perform(put(url)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isOk())
				   .andExpect(jsonPath("$.data.id").value(request.id().toString()))
				   .andExpect(jsonPath("$.data.version").value(request.version() + 1));

			verify(surveyService, times(1)).updateSurvey(any());
		}
	}

	@DisplayName("응답 제출")
	@Nested
	class SubmitResponse {
		final String url = "/api/v1/survey";
		final UUID surveyId = UUID.randomUUID();
		final int version = 1;

		@Test
		@DisplayName("설문이 존재하지 않으면 404 status와 error를 반환한다.")
		void throwExceptionWhenSurveyNotFound () throws Exception {
			SubmitResponseRequest request = ResponseFixtures.validSubmitRequest();

			given(surveyService.submitResponse(any()))
					.willThrow(new SurveyNotFoundException());

			mockMvc.perform(post(url + "/{surveyId}/version/{version}/response", surveyId, version)
						   .content(toJson(request))
						   .contentType(MediaType.APPLICATION_JSON))
				   .andExpect(status().isNotFound())
				   .andExpect(jsonPath("$.error.code").value(ErrorCodes.SURVEY_NOT_FOUND.name()))
				   .andExpect(jsonPath("$.error.message").value(ErrorCodes.SURVEY_NOT_FOUND.message));

			verify(surveyService, times(1)).submitResponse(any());
		}

		@Test
		@DisplayName("필수 질문에 대한 응답이 없으면 400 status와 error를 반환한다.")
		void throwExceptionWhenAnswerIsNullForRequiredQuestion () {
			throw new NotImplementedTestException();
		}

		@Test
		@DisplayName("설문 조사에 등록된 필수 질문이 누락된 경우 400 status와 error를 반환한다.")
		void throwExceptionWhenExactlyAnswersForRequiredQuestion () {
			throw new NotImplementedTestException();
		}

		@Test
		@DisplayName("단일 선택 리스트형 질문에 응답이 2개면 400 status와 error를 반환한다.")
		void throwExceptionWhenNotMatchedAnswersForSingleSelectQuestion () {
			throw new NotImplementedTestException();
		}

		@Test
		@DisplayName("중복 응답을 제출한 경우 400 status와 error를 반환한다.")
		void throwExceptionWhenDuplicateResponse () {
			throw new NotImplementedTestException();
		}

		@Test
		@DisplayName("응답 정보가 올바른 경우 200 status와 data를 반환한다.")
		void shouldSumbitResponse () {
			throw new NotImplementedTestException();
		}
	}

	@DisplayName("응답 조회")
	@Nested
	class GetResponse {
		@Test
		@DisplayName("요청한 설문 조사의 응답이 없으면 404 status와 error를 반환한다.")
		void throwExceptionWhenResponseNotFoundBySurveyInfo () {
			throw new NotImplementedTestException();
		}

		@Test
		@DisplayName("설문 조사에 대한 응답이 존재하면 200 status와 data를 반환한다.")
		void shouldGetResponsesBySurveyInfo () {
			throw new NotImplementedTestException();
		}
	}

	private String toJson (Object value) throws JsonProcessingException {
		return objectMapper.writeValueAsString(value);
	}
}

File: project/kyungrok-onboarding/src/test/kotlin/kr/innercircle/onboarding/survey/unit/controller/SurveyControllerTest.kt
package kr.innercircle.onboarding.survey.unit.controller

import com.fasterxml.jackson.databind.ObjectMapper
import kr.innercircle.onboarding.survey.controller.SurveyController
import kr.innercircle.onboarding.survey.domain.Survey
import kr.innercircle.onboarding.survey.domain.SurveyItemInputType
import kr.innercircle.onboarding.survey.dto.request.CreateSurveyItemOptionRequest
import kr.innercircle.onboarding.survey.dto.request.CreateSurveyItemRequest
import kr.innercircle.onboarding.survey.dto.request.CreateSurveyRequest
import kr.innercircle.onboarding.survey.service.SurveyService
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.any
import org.mockito.BDDMockito.given
import org.mockito.BDDMockito.verify
import org.mockito.Mockito.never
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*

/**
 * packageName : kr.innercircle.onboarding.survey.unit.controller
 * fileName    : SurveyControllerTest
 * author      : ckr
 * date        : 25. 6. 21.
 * description :
 */

@ExtendWith(SpringExtension::class)
@WebMvcTest(SurveyController::class)
class SurveyControllerTest {
    private val objectMapper = ObjectMapper()

    @Autowired
    private lateinit var mockMvc: MockMvc

    @MockitoBean
    lateinit var surveyService: SurveyService

    @Test
    fun 정상적인_설문조사_생성_요청시_201_상태코드와_성공_메시지를_반환한다() {
        // Given
        val validRequest = createValidSurveyRequest()
        val createdSurvey = Survey(
            id = 1L,
            name = "테스트 설문조사",
            description = "테스트 설명"
        )

        given(surveyService.createSurvey(validRequest)).willReturn(createdSurvey)

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(validRequest))
        )
            .andExpect(status().isCreated)
            .andExpect(content().contentType(MediaType.APPLICATION_JSON))
            .andExpect(jsonPath("$.success").value(true))
            .andExpect(jsonPath("$.message").value("설문조사가 생성되었습니다."))
            .andExpect(jsonPath("$.data").doesNotExist())
            .andExpect(jsonPath("$.error").doesNotExist())

        verify(surveyService).createSurvey(validRequest)
    }

    @Test
    fun 설문조사_이름이_비어있으면_400_상태코드를_반환한다() {
        // Given
        val invalidRequest = createValidSurveyRequest().copy(name = "")

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(invalidRequest))
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(invalidRequest)
    }

    @Test
    @DisplayName("설문조사 이름이 null이면 400 상태코드를 반환한다")
    fun `should return 400 when survey name is null`() {
        // Given
        val invalidRequestJson = """
            {
                "name": null,
                "description": "테스트 설명",
                "surveyItems": [
                    {
                        "name": "질문1",
                        "description": "질문1 설명",
                        "inputType": "SHORT_TEXT",
                        "options": [{"option": "옵션1"}],
                        "isRequired": true
                    }
                ]
            }
        """.trimIndent()

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(invalidRequestJson)
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(any())
    }

    @Test
    @DisplayName("설문조사 항목이 비어있으면 400 상태코드를 반환한다")
    fun `should return 400 when survey items is empty`() {
        // Given
        val invalidRequest = createValidSurveyRequest().copy(surveyItems = emptyList())

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(invalidRequest))
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(invalidRequest)
    }

    @Test
    @DisplayName("설문조사 항목이 10개를 초과하면 400 상태코드를 반환한다")
    fun `should return 400 when survey items exceed maximum size`() {
        // Given
        val tooManyItems = (1..11).map { createValidSurveyItemRequest("질문$it") }
        val invalidRequest = createValidSurveyRequest().copy(surveyItems = tooManyItems)

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(invalidRequest))
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(invalidRequest)
    }

    @Test
    @DisplayName("설문조사 항목이 null이면 400 상태코드를 반환한다")
    fun `should return 400 when survey items is null`() {
        // Given
        val invalidRequestJson = """
            {
                "name": "테스트 설문조사",
                "description": "테스트 설명",
                "surveyItems": null
            }
        """.trimIndent()

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(invalidRequestJson)
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(any())
    }

    @Test
    @DisplayName("설문조사 항목 이름이 비어있으면 400 상태코드를 반환한다")
    fun `should return 400 when survey item name is blank`() {
        // Given
        val invalidSurveyItem = createValidSurveyItemRequest("").copy(name = "")
        val invalidRequest = createValidSurveyRequest().copy(surveyItems = listOf(invalidSurveyItem))

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(invalidRequest))
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(invalidRequest)
    }

    @Test
    @DisplayName("설문조사 항목의 입력 타입이 null이면 400 상태코드를 반환한다")
    fun `should return 400 when survey item input type is null`() {
        // Given
        val invalidRequestJson = """
            {
                "name": "테스트 설문조사",
                "description": "테스트 설명",
                "surveyItems": [
                    {
                        "name": "질문1",
                        "description": "질문1 설명",
                        "inputType": null,
                        "options": [{"option": "옵션1"}],
                        "isRequired": true
                    }
                ]
            }
        """.trimIndent()

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(invalidRequestJson)
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(any())
    }

    @Test
    @DisplayName("설문조사 항목의 옵션이 null이면 400 상태코드를 반환한다")
    fun `should return 400 when survey item options is null`() {
        // Given
        val invalidRequestJson = """
            {
                "name": "테스트 설문조사",
                "description": "테스트 설명",
                "surveyItems": [
                    {
                        "name": "질문1",
                        "description": "질문1 설명",
                        "inputType": "SHORT_TEXT",
                        "options": null,
                        "isRequired": true
                    }
                ]
            }
        """.trimIndent()

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(invalidRequestJson)
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(any())
    }

    @Test
    @DisplayName("설문조사 항목 옵션이 비어있으면 400 상태코드를 반환한다")
    fun `should return 400 when survey item option is blank`() {
        // Given
        val invalidOption = CreateSurveyItemOptionRequest("")
        val invalidSurveyItem = createValidSurveyItemRequest("질문1").copy(options = listOf(invalidOption))
        val invalidRequest = createValidSurveyRequest().copy(surveyItems = listOf(invalidSurveyItem))

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(invalidRequest))
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(invalidRequest)
    }

    @Test
    @DisplayName("설문조사 항목의 필수 여부가 null이면 400 상태코드를 반환한다")
    fun `should return 400 when survey item isRequired is null`() {
        // Given
        val invalidRequestJson = """
            {
                "name": "테스트 설문조사",
                "description": "테스트 설명",
                "surveyItems": [
                    {
                        "name": "질문1",
                        "description": "질문1 설명",
                        "inputType": "SHORT_TEXT",
                        "options": [{"option": "옵션1"}],
                        "isRequired": null
                    }
                ]
            }
        """.trimIndent()

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(invalidRequestJson)
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(any())
    }

    @Test
    @DisplayName("잘못된 JSON 형식이면 400 상태코드를 반환한다")
    fun `should return 400 when request body has invalid JSON format`() {
        // Given
        val invalidJson = """
            {
                "name": "테스트 설문조사",
                "description": "테스트 설명",
                "surveyItems": [
                    {
                        "name": "질문1",
                        "description": "질문1 설명",
                        "inputType": "SHORT_TEXT",
                        "options": [{"option": "옵션1"}],
                        "isRequired": true
                    }
                // 누락된 닫는 괄호
            }
        """.trimIndent()

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(invalidJson)
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(any())
    }

    @Test
    @DisplayName("Content-Type이 application/json이 아니면 415 상태코드를 반환한다")
    fun `should return 415 when content type is not application json`() {
        // Given
        val validRequest = createValidSurveyRequest()

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_XML)
                .content(objectMapper.writeValueAsString(validRequest))
        )
            .andExpect(status().isUnsupportedMediaType)

        verify(surveyService, never()).createSurvey(any())
    }

    @Test
    @DisplayName("요청 본문이 비어있으면 400 상태코드를 반환한다")
    fun `should return 400 when request body is empty`() {
        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content("")
        )
            .andExpect(status().isBadRequest)

        verify(surveyService, never()).createSurvey(any())
    }

    @Test
    @DisplayName("서비스에서 예외가 발생하면 예외가 전파된다")
    fun `should propagate exception when service throws exception`() {
        // Given
        val validRequest = createValidSurveyRequest()
        given(surveyService.createSurvey(validRequest))
            .willThrow(RuntimeException("서비스 오류"))

        // When & Then
        try {
            mockMvc.perform(
                post("/surveys")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(validRequest))
            )
        } catch (e: Exception) {
            // 예외가 발생하는 것이 정상
        }

        verify(surveyService).createSurvey(validRequest)
    }

    @Test
    @DisplayName("모든 입력 타입에 대해 정상적으로 처리된다")
    fun `should handle all input types successfully`() {
        // Given
        val inputTypes = SurveyItemInputType.entries
        inputTypes.forEach { inputType ->
            val request = createValidSurveyRequest().copy(
                surveyItems = listOf(
                    createValidSurveyItemRequest("질문").copy(inputType = inputType)
                )
            )
            val createdSurvey = Survey(
                id = 1L,
                name = "테스트 설문조사",
                description = "테스트 설명"
            )

            given(surveyService.createSurvey(request)).willReturn(createdSurvey)

            // When & Then
            mockMvc.perform(
                post("/surveys")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(request))
            )
                .andExpect(status().isCreated)
                .andExpect(jsonPath("$.success").value(true))
                .andExpect(jsonPath("$.message").value("설문조사가 생성되었습니다."))

            verify(surveyService).createSurvey(request)
        }
    }

    @Test
    @DisplayName("description이 null이어도 정상적으로 처리된다")
    fun `should handle null description successfully`() {
        // Given
        val requestWithNullDescription = createValidSurveyRequest().copy(description = null)
        val createdSurvey = Survey(
            id = 1L,
            name = "테스트 설문조사",
            description = null
        )

        given(surveyService.createSurvey(requestWithNullDescription)).willReturn(createdSurvey)

        // When & Then
        mockMvc.perform(
            post("/surveys")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(requestWithNullDescription))
        )
            .andExpect(status().isCreated)
            .andExpect(jsonPath("$.success").value(true))
            .andExpect(jsonPath("$.message").value("설문조사가 생성되었습니다."))

        verify(surveyService).createSurvey(requestWithNullDescription)
    }

    private fun createValidSurveyRequest(): CreateSurveyRequest {
        return CreateSurveyRequest(
            name = "테스트 설문조사",
            description = "테스트 설명",
            surveyItems = listOf(createValidSurveyItemRequest("질문1"))
        )
    }

    private fun createValidSurveyItemRequest(name: String): CreateSurveyItemRequest {
        return CreateSurveyItemRequest(
            name = name,
            description = "질문 설명",
            inputType = SurveyItemInputType.SHORT_TEXT,
            options = listOf(CreateSurveyItemOptionRequest("옵션1")),
            isRequired = true
        )
    }
}

